### PR TITLE
WND  data preprocess with feature table

### PR DESCRIPTION
**Pull request by jenniew**
WND  data preprocess with feature table.

### CODE REVIEW

File: zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala [hunk 1]
```diff
@@ -78,4 +78,8 @@ private[friesian] object Utils {
       Array(Tuple2(part_id, rows.size)).iterator
     }
   }
+
+  def hashBucket(content: Any, bucketSize: Int = 1000, start: Int = 0): Int = {
+    return (content.hashCode() % bucketSize + bucketSize) % bucketSize + start
```
Comment (jason-dai) [inline zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala:83]: why adding `bucketSize` here?

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    elif args.cluster_mode == "yarn":
+        init_orca_context("yarn-client", cores=args.executor_cores,
+                          num_nodes=args.num_executor, memory=args.executor_memory,
+                          driver_cores=args.driver_cores, driver_memory=args.driver_memory,
+                          conf=conf)
+    time_start = time()
+    paths = [os.path.join(args.input_folder, 'day_%d.parquet' % i) for i in args.day_range]
+    tbl = FeatureTable.read_parquet(paths)
+    idx_list = tbl.gen_string_idx(CAT_COLS, freq_limit=args.frequency_limit)
+    item_size = idx_list[1].count()
```
Comment (jason-dai) [inline pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:101]: `idx_list[1].size()`

And where is `item_size` used?

File: pyzoo/zoo/friesian/feature/table.py [hunk 3]
```diff
@@ -118,6 +125,27 @@ def gen_string_idx(self, columns, freq_limit):
     def _clone(self, df):
         return FeatureTable(df)
 
+    def cross_columns(self, cross_column_list, cross_sizes):
```
Comment (jason-dai) [inline pyzoo/zoo/friesian/feature/table.py:128]:
```python
def cross_columns(self, crossed_column, bucket_size)
```
`bucket_size` can be an Int value, in this case `crossed_column` is a list of columns names (i.e., similar to `tf.feature_column`)

Or `bucket_size` can be a list of Int values, and `crossed_column` is a list of list (of columns names), and their length should match.

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    tbl_all_data = tbl.encode_string(CAT_COLS, idx_list)\
+        .fillna(0, INT_COLS + CAT_COLS)
+    for col in INT_COLS:
+        tbl_all_data = tbl_all_data.normalize(col)
```
Comment (jason-dai) [inline pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:105]: `tbl_all_data.normalize(INT_COLS)`?

File: zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala [hunk 5]
```diff
@@ -78,4 +78,8 @@ private[friesian] object Utils {
       Array(Tuple2(part_id, rows.size)).iterator
     }
   }
+
+  def hashBucket(content: Any, bucketSize: Int = 1000, start: Int = 0): Int = {
```
Comment (hkvision) [inline zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala:82]: Can we directly use buckBucket in Utils under zoo/models/recommendation?

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 2 above)
Comment (jenniew) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:101]: no need for item_size. Removed

File: zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala (hunk 1 above)
Comment (jenniew) [reply zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala:83]: content.hashCode() % bucketSize may be negative, adding bucketSize to make it positive.

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 4 above)
Comment (jenniew) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:105]: updated.

File: zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala (hunk 5 above)
Comment (jenniew) [reply zoo/src/main/scala/com/intel/analytics/zoo/friesian/feature/Utils.scala:82]: buckBucket can only support string column. crossColumn may support any type columns.

File: pyzoo/zoo/friesian/feature/table.py (hunk 3 above)
Comment (jenniew) [reply pyzoo/zoo/friesian/feature/table.py]: changed to 
``` 
def cross_columns(self, crossed_columns, bucket_sizes)
```
where `crossed_columns` is a list of list, and `bucket_sizes` is a list of int values.

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py [hunk 6]
```diff
@@ -69,20 +73,39 @@ def _parse_args():
...
+
     return args
 
 
 if __name__ == '__main__':
     args = _parse_args()
     if args.cluster_mode == "local":
-        init_orca_context("local", cores=args.executor_cores, memory=args.executor_memory)
+        # init_orca_context("local", cores=args.executor_cores, memory=args.executor_memory)
+        sc = init_nncontext("wnd")
```
Comment (jason-dai) [inline pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:108]: why `ini_nncontext`?

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py [hunk 7]
```diff
@@ -69,20 +73,39 @@ def _parse_args():
...
+        for fl in frequency_limit:
+            frequency_pair = fl.split(":")
+            if len(frequency_pair) == 1:
+                default_limit = int(frequency_pair[0])
+            elif len(frequency_pair) == 2:
+                frequency_limit_dict[frequency_pair[0]] = frequency_pair[1]
+    if len(frequency_limit_dict) > 0:
+        args.frequency_limit = frequency_limit_dict
+    else:
+        args.frequency_limit = default_limit
```
Comment (jason-dai) [inline pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:99]: Why do you need to parse `frequency_limit` here? Why not just pass it to `gen_string_idx` directly?

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+    return args
+
+
+if __name__ == '__main__':
+    args = _parse_args()
+    if args.cluster_mode == "local":
+        # init_orca_context("local", cores=args.executor_cores, memory=args.executor_memory)
+        sc = init_nncontext("wnd")
+    elif args.cluster_mode == "standalone":
```
Comment (jason-dai) [inline pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:109]: Why do we want to use standalone here? The current "standalone" implementation still have some constraints @hkvision

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                          conf=conf)
+
+    time_start = time()
+    paths = [os.path.join(args.input_folder, 'sample_test_day_%d.parquet' % i) for i in args.day_range]
+    tbl = FeatureTable.read_parquet(paths)
+    # change name for all columns
+    columns = dict([("_c{}".format(i), "c{}".format(i)) for i in range(40)])
+    tbl = tbl.rename(columns)
+    idx_list = tbl.gen_string_idx(CAT_COLS, freq_limit=args.frequency_limit)
+    cat_sizes = [idx.count() for idx in idx_list]
```
Comment (jason-dai) [inline pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:128]: change to `idx.size()`?

**Comment by jason-dai** [comment]
Is the PR ready for merge?

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 6 above)
Comment (jenniew) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py]: removed

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 7 above)
Comment (jenniew) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:101]: the input frequency_limit is string, we need to parse to get dict and pass to gen_string_idx

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 9 above)
Comment (jenniew) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:131]: no size() function for idx now. Do we need to add this function?

**Comment by jenniew** [comment]
> Is the PR ready for merge?

I just optimize normalization and test the performance. The whole preprocess time is decreased to half an hour. I think it may be ready.

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 7 above)
Comment (jason-dai) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:101]: we can simply use Int for the example

File: pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py (hunk 9 above)
Comment (jason-dai) [reply pyzoo/zoo/examples/friesian/feature/wnd_preprocessing.py:131]: Yes, add `size`

File: pyzoo/zoo/friesian/feature/table.py [hunk 10]
```diff
@@ -489,3 +533,10 @@ def write_parquet(self, path, mode="overwrite"):
         """
         path = path + "/" + self.col_name + ".parquet"
         self.df.write.parquet(path, mode=mode)
+
+    def size(self):
+        """
+        Size of the StringIndex
+        :return:
+        """
+        return self.df.select(self.col_name, "id").distinct().count()
```
Comment (jason-dai) [inline pyzoo/zoo/friesian/feature/table.py:542]: why not simply `df.count`?

Comment (jenniew) [reply pyzoo/zoo/friesian/feature/table.py]: updated.

File: pyzoo/zoo/friesian/feature/table.py [hunk 11]
```diff
@@ -489,3 +545,10 @@ def write_parquet(self, path, mode="overwrite"):
         """
         path = path + "/" + self.col_name + ".parquet"
         self.df.write.parquet(path, mode=mode)
+
+    def size(self):
```
Comment (jason-dai) [inline pyzoo/zoo/friesian/feature/table.py:549]: maybe you should just change `Table.count()` to `Table.size()`?

**Comment by jenniew** [comment]
Jenkins pass: http://10.239.47.210:18888/job/ZOO-PR-Validation/5531/

**Pull request merged**